src/lib/react/useLocalMedia.ts

```diff
@@ -1,25 +1,119 @@
-import { useEffect, useState } from "react";
+import { useEffect, useReducer, useState } from "react";
+import LocalMedia from "../LocalMedia";
 
-export default function useLocalMedia() {
-    const [localStream, setLocalStream] = useState<MediaStream | undefined>(undefined);
+interface LocalMediaState {
+    currentCameraDeviceId?: string;
+    currentMicrophoneDeviceId?: string;
+    cameraDevices: MediaDeviceInfo[];
+    localStream?: MediaStream;
+    microphoneDevices: MediaDeviceInfo[];
+    speakerDevices: MediaDeviceInfo[];
+}
+
+const initialState: LocalMediaState = {
+    cameraDevices: [],
+    microphoneDevices: [],
+    speakerDevices: [],
+};
+
+interface LocalMediaActions {
+    setCameraDevice: InstanceType<typeof LocalMedia>["setCameraDevice"];
+    setMicrophoneDevice: InstanceType<typeof LocalMedia>["setMicrophoneDevice"];
+    toggleCameraEnabled: InstanceType<typeof LocalMedia>["toggleCameraEnabled"];
+    toggleMicrophoneEnabled: InstanceType<typeof LocalMedia>["toggleMichrophoneEnabled"];
+}
+
+type LocalMediaEvents =
+    | {
+          type: "DEVICE_LIST_UPDATED";
+          payload: {
+              cameraDevices: MediaDeviceInfo[];
+              microphoneDevices: MediaDeviceInfo[];
+              speakerDevices: MediaDeviceInfo[];
+          };
+      }
+    | {
+          type: "LOCAL_STREAM_UPDATED";
+          payload: { stream: MediaStream; currentCameraDeviceId?: string; currentMicrophoneDeviceId?: string };
+      };
+
+export type LocalMediaRef = { state: LocalMediaState; actions: LocalMediaActions; _ref: LocalMedia | null };
+
+function reducer(state: LocalMediaState, action: LocalMediaEvents): LocalMediaState {
+    switch (action.type) {
+        case "DEVICE_LIST_UPDATED":
+            return {
+                ...state,
+                ...action.payload,
+            };
+        case "LOCAL_STREAM_UPDATED":
+            return {
+                ...state,
+                currentCameraDeviceId: action.payload.currentCameraDeviceId,
+                currentMicrophoneDeviceId: action.payload.currentMicrophoneDeviceId,
+                localStream: action.payload.stream,
+            };
+        default:
+            return state;
+    }
+}
+
+export default function useLocalMedia(
+    constraints: MediaStreamConstraints = { audio: true, video: true }
+): LocalMediaRef {
+    const [localMedia, setLocalMedia] = useState<LocalMedia | null>(null);
```

Comment:
Why not set the initial `localMedia` here instead of in the useEffect?

Reply:
Good question. I tried doing that with the RoomConnection, but that ended up creating new instances multiple times.

I can have another go at cleaning this up 👍

Reply:
Hah! Got it to work. The key was to wrap the `new LocalMedia(...)` call in an anonymous function. Found the tip here https://stackoverflow.com/a/64131447. This makes everything much cleaner I agree, as we can always assume there is an instance 👍